consensus/src/quorum_store/batch_generator.rs

```diff
@@ -146,16 +146,38 @@ impl BatchGenerator {
         num_txns_in_bucket: usize,
         expiry_time: u64,
         bucket_start: u64,
+        total_batches_remaining: &mut u64,
+        total_txns_remaining: &mut u64,
+        total_bytes_remaining: &mut u64,
     ) -> bool {
         let mut remaining_txns = num_txns_in_bucket;
         while remaining_txns > 0 {
-            if batches.len() == self.config.sender_max_num_batches {
+            if *total_batches_remaining == 0
+                || *total_txns_remaining == 0
+                || *total_bytes_remaining == 0
+            {
                 return false;
             }
-            let num_batch_txns = std::cmp::min(self.config.sender_max_batch_txns, remaining_txns);
+            let num_take_txns = std::cmp::min(self.config.sender_max_batch_txns, remaining_txns);
+            let mut batch_bytes: u64 = 0;
+            let num_batch_txns = txns
+                .iter()
+                .take(num_take_txns)
+                .take_while(|txn| {
+                    let txn_bytes = txn.txn_bytes_len() as u64;
+                    batch_bytes += txn_bytes;
+                    *total_txns_remaining = total_txns_remaining.saturating_sub(1);
+                    *total_bytes_remaining = total_bytes_remaining.saturating_sub(txn_bytes);
```

Comment:
isn't the saturating_sub able to exceed the limit? if the remaining is 1 bytes but txn_bytes is 64kb?

I thought we should do checked_sub and short circuit on errors

Reply:
Hmm, actually for the case you mention this works because the check below is for `*total_bytes_remaining > 0`. But I do think there's an off-by-one error here where if the max is 64 kb and the txn is 64 bytes it gets rejected. Let me change it

Reply:
isn't it simpler to just to checked_sub and not checking at the end? anyway approving to unblock